mirgecom/logging_quantities.py

```diff
@@ -0,0 +1,321 @@
+"""Support for time series logging."""
+
+__copyright__ = """
+Copyright (C) 2020 University of Illinois Board of Trustees
+"""
+
+__license__ = """
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in
+all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+"""
+
+__doc__ = """
+.. autoclass:: StateConsumer
+.. autoclass:: DiscretizationBasedQuantity
+.. autoclass:: KernelProfile
+.. autoclass:: PythonMemoryUsage
+.. autofunction:: initialize_logmgr
+.. autofunction:: logmgr_add_device_name
+.. autofunction:: logmgr_add_default_discretization_quantities
+.. autofunction:: add_package_versions
+.. autofunction:: set_sim_state
+"""
+
+from logpyle import (LogQuantity, LogManager, MultiLogQuantity, add_run_info,
+    add_general_quantities, add_simulation_quantities)
+from meshmode.array_context import PyOpenCLArrayContext
+from meshmode.discretization import Discretization
+import pyopencl as cl
+
+from typing import Optional, Callable
+import numpy as np
+
+
+def initialize_logmgr(enable_logmgr: bool,
+                      filename: str = None, mode: str = "wu",
+                      mpi_comm=None) -> LogManager:
+    """Create and initialize a mirgecom-specific :class:`logpyle.LogManager`."""
+    if not enable_logmgr:
+        return None
+
+    logmgr = LogManager(filename=filename, mode=mode, mpi_comm=mpi_comm)
+
+    add_run_info(logmgr)
+    add_package_versions(logmgr)
+    add_general_quantities(logmgr)
+    add_simulation_quantities(logmgr)
+
+    try:
+        logmgr.add_quantity(PythonMemoryUsage())
+    except ImportError:
+        from warnings import warn
+        warn("psutil module not found, not tracking memory consumption."
+             "Install it with 'pip install psutil'")
+
+    return logmgr
+
+
+def logmgr_add_device_name(logmgr: LogManager, queue: cl.CommandQueue):
+    """Add the OpenCL device name to the log."""
+    logmgr.set_constant("cl_device_name", str(queue.device))
+
+
+def logmgr_add_default_discretization_quantities(logmgr: LogManager, discr, dim,
```

Comment:
```suggestion
def logmgr_add_all_discretization_quantities(logmgr: LogManager, discr, dim,
```

Reply:
The quantities added here aren't necessarily all discretization related quantities, so renaming this is maybe confusing.

Reply:
Agreed. `add_many`?

Reply:
Could be, or `add_basic`?

Reply:
Don't like `add_basic` so much.

Reply:
renamed to `add_many`.